examples/hosts/app-integration/live-schema-upgrade/src/modelVersion2/containerCode.ts

```diff
@@ -0,0 +1,137 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ModelContainerRuntimeFactory } from "@fluid-example/example-utils";
+import { IContainer, IFluidCodeDetails } from "@fluidframework/container-definitions";
+import { ConnectionState } from "@fluidframework/container-loader";
+import { IContainerRuntime } from "@fluidframework/container-runtime-definitions";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+
+import { DiceRollerInstantiationFactory, IDiceRoller } from "./diceRoller";
+import { DiceCounterInstantiationFactory, IDiceCounter } from "./diceCounter";
+
+const diceRollerId = "dice-roller";
+const diceCounterId = "dice-counter";
+
+/**
+ * The data model for our application.
+ *
+ * @remarks Note that this version of the model has a dice counter object in addition to the dice roller object.
+ */
+export interface IDiceRollerAppModel {
+	/**
+	 * DiceRoller data object to track the current dice roll.
+	 */
+	readonly diceRoller: IDiceRoller;
+
+	/**
+	 * DiceCounter data object to track the number of times the dice have been rolled.
+	 */
+	readonly diceCounter: IDiceCounter;
+
+	/**
+	 * Returns the current version of the container.
+	 */
+	readonly getCurrentVersion: () => string;
+
+	/**
+	 * Perform the code proposal to upgrade the container to the latest version.
+	 */
+	readonly upgrade: (targetVersion: string) => Promise<void>;
+}
+
+class DiceRollerAppModel implements IDiceRollerAppModel {
+	public constructor(
+		public readonly diceRoller: IDiceRoller,
+		public readonly diceCounter: IDiceCounter,
+		public readonly container: IContainer,
+	) {
+		container.on("closed", () => {
+			// Ensure the user can't roll the dice after the container is closed.
+			diceRoller.close();
+		});
+	}
+
+	public getCurrentVersion() {
+		return this.container.getSpecifiedCodeDetails()?.package as string;
+	}
+
+	public async upgrade(targetVersion: string) {
+		const currentVersion = this.getCurrentVersion();
+		if (currentVersion === targetVersion) {
+			// We shouldn't try to upgrade if we are already on the latest version.
+			return;
+		}
+		console.log(`Upgrading to ${targetVersion}`);
+		if (this.container.connectionState !== ConnectionState.Connected) {
+			await new Promise((resolve) => {
+				this.container.once("connected", resolve);
+			});
+		}
+		const proposal: IFluidCodeDetails = { package: targetVersion };
+		this.container
+			.proposeCodeDetails(proposal)
```

Comment:
Glancing at the code, this will fail if disconnect happens. Is this Ok? Feels like either this code, or Container's implementation should deal with proper retries (however we define them).

Reply:
Added retry logic in latest :)